tests/virt_autotest/guest_installation_run.pm

```diff
@@ -53,6 +53,16 @@ sub analyzeResult {
     return $result;
 }
 
+sub post_execute_script_assertion {
```

Comment:
Why this function?

Reply:
I already explained in the description. It lets you to have more customized way to do your own assertion further if you can not do that in post_execute_script_run. And it help keep the current test logic for all test projects and still keeps them tidy.

Reply:
You said you know my design, but actually I think you do not. Because you feel lost and call all them unplanned changes, so I have to say you did not even give the description a careful reading before commenting all these kind of absurd and misleading comments. 
All these new functions are to facilitate post_fail_hook transformation. post_fail_hook introduces a new path in the test flow if test failed so it needs all test modules to switch path if they failed. This is very different to what we do currently. Currently you can do whatever you want in your own test module and even do monolithic way of development, yes, I starts to seeing that in many test modules, for example, guest upgrdae/migration. This monolithic way can lead to chaos, especially with post_fail_hook in pace, because we need to design more carefully about what needs to be done before going into post_fail_hook and what should happen after in post_fail_hook. The same king of thing also applies to other parts in virt_autotest_base module, for example, what needs to be done after executing script immediately and what should be handled before generating junit log. And at the same time, many test modules have similar or different ways to do all these, in other words, they scattered everywhere without even being noticed if without careful look. So modular way of doing development comes into play. 
I do not want to give new introductions of these functions again here, you should need to read them carefully in the description again. Most of them is to re-organize stuff in individual test modules. I think only post_execute_script_assertion has new things in it, because I need to add new assertion for guest installation and consolidate existing ones for guest upgrade and migration. It is very important. It help handle failed path of test efficiently and preserved existing test logic and still make them look tidy.
Obviously my modular design is much better. Modular design always bring benefits, for example, encouraging further innovation. On the contrary, monolithic way hurdles even possibility of further extension.
All these are very easy, and it should be with modular design. Please do not prevent yourself from getting familiar with newer better ideas and the excuse is they are hard to understand. This not true in every aspect.
Is there risk ? Yes, every new way of doing things come with risk. We remove risk by doing test and enough test. This is easy and I have already done that.

Reply:
Thanks for the  comment. I agree from some aspects, while not for some other aspects. I won't argue for this now.

To clear misunderstanding here, I want to rephrase my points. I mainly had 2 concerns when I gave the previous comment:
1) No enough time for careful enough validation? : 
The PR changes all traditional prj1-prj6. I am afraid in last 2-weeks sprint, you might only have 1 week for validation. It was not likely to be long  enough for careful enough validation with full run and various pass/fail/timeout/incomplete/etc job result situations.
And at the current time being, for which next step is to start sp3 preparation during which we won't have time to fix much potential issues brought by this, so it is RISK point I see for project.  // I guess you know it too. If it's merged and really brought issues, you will have to be responsible for them(6 projects), that is 
 to spare time for fixing them and not block other's tasks in roadmap while finishing your tasks in the roadmap. I am afraid it is hard for you to manage all these parally. But you commented that you had done enough validation for this. If yes, the risk is low and this concern can be cleared now.

2) Can the benefit from this major refinement for virt_autotest_base win the impacts that it bring? : I said I got lost on it. I used "got", rather than "get" in original comment. So yes, I did get lost at the beginning, after I read through it, I understood it. So now I understand it, do not worry about it. However from the new function names, it is not easy to remember/tell directly what they do and will need to refer to whole calling process to recall who is what.  I guess others will repeat the similar process to get used to it too. Whether this is a good design is another topic, which let's discuss  after we confirm that it is necessary .

This RR from my POV introduces mainly two things: 
a) upload guest/host  logs: this is what we planned 
b) the other side changes to make it work as expected and work well: this part can have different way to do. 
The way picked in this PR is baseclass major/fatal refinement way. It brings the above  comprehensive validation task + major effort to get familiar with new base class design task, which is unexpected. If some minor way is taken, very limited validation will be needed, and we can even ignore other member's effort to get familiar with new base class design. Why is  this major reconstruction necessary? Why can not mainly simply add "upload guest/host logs "directly as feature test base class does(of course may need some other minor necessary changes)? I am sorry I do not see it.

Finally, I am willing to accept it, IF
- there are reasons that make it necessary to take this major baseclass refinement
- or, even if not necessary, but you already made it(comprehensive validations are done), and all other members can accept the major change and are willing to(as well as have time to) get familiar with the changes
- or, if later after merged, during sp3 preparations, untested issues exposed on various prj1-prj6 tests, but you have time to fix them and make others not blocked, while delivering your other tasks in roadmap(hardly can be delayed)

Otherwise, I may suggest minor change way for b).

Reply:
Besides, please let me apologize for giving the very first comment in this conversation before reading thru the whole design. I should have reverted it after I realized that I gave too early after I read more of the whole changes. I jumped into a meeting not long after this first comment. When I came back to this PR, wayne already replied to it. Sorry for this.

Reply:
1. This is a not very small change from the beginning, because there is no post_fail_hook in virt_autotest_base module before. This kind of change requires you to do more comprehensive changes whatever you may think of it as. I know this point. So I am prepared to take the responsibility and the risk. So I can manage it. And I think everyone should know that, or at least, know this is a shift instead of just piling up code
2. This is a change to base class. Base class itself is a common place to handle common things for all individual test modules, for example, post_fail_hook. And implementing post_fail_hook in base class saves duplicate efforts and free up individuals to invent the real tests. So this is why it should be in base class. If you understand this point you should also understand what occurs to other parts of this implementation. Base case becomes the control center of common things, including work flow and do judgement by test results. This requires clearer test structure and flow instead of putting everything into one function and at the same time adding arbitrary code into test modules themselves. The former is modular and the latter is monolithic. The latter can lead to broken test flow very easily if not handled correctly, but the former simplifies things and make everything clearer. So new functions appear to implement necessary changes and also new functions appear to wrap up scattered code. All these can help simplify test flow, make it more clear to everyone. If you look at the sequence of called functions in run_test in base class, you should know what I mean. It is much simpler than before instead of complicating things and costing everyone extra efforts to take in new things. That is not the case. It is just the right thing to do and to adopt.
3. As I already said a little bit before, this is the step must be taken to move forward, otherwise I can not accept it as a complete solution. Just imaging that situation, a standalone post_fail_hook without any other changes, frightens me. If this was the case, to be honest, I even do not how to fix it if something unexpected happens, because there are too many scattered code performs various functionalities. A kind of chaos situation to me.
4. What I am going to do or what I have already done is to make sure all of them works. So I have no worry to burden myself or prevent myself from doing other things or rushing to fix something. I will just make them work as expected. This is part of job. I have no questions.
5. I think there should be no more arguments about what is this all about or is this good ? It is better obviously. Anyone with some knowledge about development should know what modular design can bring to everyone and what monolithic design means.

Reply:
I am not just doing fancy things or do what I like to do. This is required by making the post_fail_hook change, because it starts the journey of transforming to modular design. Actually we do use modular design always, otherwise there should be no openQA APIs, no various openQA services, no various modules. But, at some point, someone may off the track or something may prevent he or she to do so, so it should be correct as well. (This is a useful article about benefits of modular design)[http://gwentechembedded.com/the-advantages-of-modular-software-and-programming/#], I think you should read to make you feel are more willing to accept.

Reply:
I fully agree with the advantages of modular design. The main problem I see is the timing to apply this modular change for b). Do it before sp3 preparation? It is not a good time I think. I know it is not small change from the beginning. But this touch deep into every existing project's "core" openqa code in all xxx_run.pm on non-trivial parts, is unexpected and risky from my POV.

Anyway, let's wait for the project owner's opinion. They are the final test owners, that accepts and are responsible for the risk. As I gave earlier, if they agreed, and you gave enough validation to eliminate the risk, I am willing to merge it, after minor potential modifications by further comments given based on accepting the main change structure.

@Julie-CAO @guoxuguang Can you help review the PR? Especially evaluate the risk on your own responsible  test projects. We count on your input for final decisions. Thanks!

Reply:
I am prudent to say "there is risk" and "this is not small change", but actually all these are handled very well. Please do not confuse risk with managed risk, they are different. The very underlying test logic does not change at all. I know your concern, but there is not too much to worry about. Re-organize and re-arrange, even re-factor is so common in development. If you fear everything, then there is not too much we can do. I think I have time to handle all these  even before SP3 preparation. Please do not over exaggerate the overall difficulty or risk to adopt this new method, there is nothing novel. I just adopt the right way to do right thing. If dividing big to smaller ones and re-arrange them altogether again poses such huge hardness to you, then I think developers will laugh at this. You should live with it.

Reply:
Whatever @Julie-CAO @guoxuguang will say, I strongly recommend you to look at this from a fresh angle. Basically this is simple and easy instead of difficult and risky. We should do things in the right way instead of sticking to ugly outdated design forever. Now it is time to do that. Modular is beautiful, please embrace it !

Reply:
Basic review this PR, pretty cool. Grete Job. @waynechen55  also, agree with the advantages of modular design too.
As the same, understood that @waynechen55 had make more validation runs to confirm this PR for now. 

Moreover, refer to our Roadmap 3 for Virtualization-Function QA, next, we will focus on SLES15SP3 OSD job group & testsuites & automation preparation as high priority task. imho, it will take most time to made the related code changes for the currently OSD Virtualization-Function code for us. 

So, I suggested hold this PR for a while, after finished SLES15SP3 OSD job group & testsuites & automation preparation,
most OSD SLES15SP3 Virtualization-Function test run were stable, then merge this PR.

Reply:
> Basic review this PR, pretty cool. Grete Job. @waynechen55 also, agree with the advantages of modular design too.
> As the same, understood that @waynechen55 had make more validation runs to confirm this PR for now.
> 
> Moreover, refer to our Roadmap 3 for Virtualization-Function QA, next, we will focus on SLES15SP3 OSD job group & testsuites & automation preparation as high priority task. imho, it will take most time to made the related code changes for the currently OSD Virtualization-Function code for us.
> 
> So, I suggested hold this PR for a while, after finished SLES15SP3 OSD job group & testsuites & automation preparation,
> most OSD SLES15SP3 Virtualization-Function test run were stable, then merge this PR.

I think you do not need to care too much about my next upcoming tasks. This should go ahead as usual. There is no conflict from my POV. I will send signal when it is ready to be merged.